### PR TITLE
controller: Schedule one-off jobs via the scheduler

### PR DESCRIPTION
**Pull request by lmars**
This adds the concept of a "job request" which is persisted to the database, and then the scheduler fulfils job requests by starting jobs in the cluster.

This currently only affects jobs requested via `client.RunJob{Attached,Detached}`, however I have plans to change both slugbuilder and worker jobs to use this too, but I will do that in subsequent PRs.

Refs #2620.

### CODE REVIEW

File: controller/types/types.go [hunk 1]
```diff
@@ -226,8 +237,16 @@ func JobDownEvents(count int) map[JobState]int {
 	return map[JobState]int{JobStateDown: count}
 }
 
+// NewJob is DEPRECATED in favour of JobRequest
```
Comment (jvatic) [inline controller/types/types.go]: The deprecation should be noted in the new_job JSON schema.

File: controller/jobs.go [hunk 2]
```diff
@@ -231,6 +422,84 @@ func (c *controllerAPI) KillJob(ctx context.Context, w http.ResponseWriter, req
...
+	if err != nil {
+		respondWithError(w, err)
+		return
+	}
+	release := releaseData.(*ct.Release)
+	if len(req.ArtifactIDs) == 0 {
+		req.ArtifactIDs = release.ArtifactIDs
+	}
+
+	// TODO: validate
```
Comment (jvatic) [inline controller/jobs.go]: `JobRequest` needs a JSON schema (you'll likely have to update [controller/schema/schema.go schemaForType](https://github.com/flynn/flynn/blob/3c89a606c28559a5cfa0eadd989c17572e52f03e/controller/schema/schema.go#L55) for `schema.Validate(req)` to work).

**Comment by lmars** [comment]
Rebased on master and comments addressed, please review.

File: controller/artifact.go [hunk 3]
```diff
@@ -129,3 +132,25 @@ func scanArtifacts(rows *pgx.Rows) (interface{}, error) {
...
+	var list []*ct.Artifact
+	var err error
+	if ids := req.URL.Query().Get("ids"); ids != "" {
+		artifacts, err := c.artifactRepo.ListIDs(strings.Split(ids, ",")...)
+		if err != nil {
+			respondWithError(w, err)
+			return
+		}
+		for _, artifact := range artifacts {
+			list = append(list, artifact)
```
Comment (titanous) [inline controller/artifact.go]: Is this necessary? Can `list` just be assigned above?

**Comment by titanous** [comment]
Closing, this patch is now quite outdated. Hopefully it can be picked up in the future.

**Pull request closed without merging**